primitives/src/v1.rs

```diff
@@ -476,3 +479,221 @@ pub struct AvailableData {
 	/// The omitted validation data.
 	pub omitted_validation: OmittedValidationData,
 }
+
+/// A helper data-type for tracking validator-group rotations.
+#[derive(Clone, Encode, Decode)]
+#[cfg_attr(feature = "std", derive(PartialEq, Debug))]
+pub struct GroupRotationInfo<N = BlockNumber> {
+	/// The block number where the session started.
+	pub session_start_block: N,
+	/// How often groups rotate. 0 means never.
+	pub group_rotation_frequency: N,
+	/// The current block number.
+	pub now: N,
+}
+
+impl GroupRotationInfo {
+	/// Returns the index of the group needed to validate the core at the given index, assuming
+	/// the given number of cores.
+	///
+	/// `core_index` should be less than `cores`, which in turn should be less than u32::max().
+	pub fn group_for_core(&self, core_index: CoreIndex, cores: usize) -> GroupIndex {
+		if self.group_rotation_frequency == 0 { return GroupIndex(core_index.0) }
+		if cores == 0 { return GroupIndex(0) }
```

Comment:
is `cores == 0` a valid case?

Reply:
Yeah, if there are no parachains registered and no parathread cores, I think we'd have 0 cores.